pkg/dataplane/syncingress.go

```diff
@@ -0,0 +1,193 @@
+			return err
+		}
+
+		klog.InfoS("Backend successfully reloaded")
+		n.metricCollector.ConfigSuccess(hash, true)
+		n.metricCollector.IncReloadCount()
+		// TODO: Below should be another gRPC call to let controller know there was an error
+		//n.recorder.Eventf(k8s.IngressPodDetails, apiv1.EventTypeNormal, "RELOAD", "NGINX reload triggered due to a change in configuration")
+	}
+
+	isFirstSync := n.runningConfig.Equal(&ingress.Configuration{})
+	if isFirstSync {
+		// For the initial sync it always takes some time for NGINX to start listening
+		// For large configurations it might take a while so we loop and back off
+		klog.InfoS("Initial sync, sleeping for 1 second")
+		time.Sleep(1 * time.Second)
+	}
+
+	retry := wait.Backoff{
+		Steps:    1 + n.cfg.DynamicConfigurationRetries,
+		Duration: time.Second,
+		Factor:   1.3,
+		Jitter:   0.1,
+	}
+
+	retriesRemaining := retry.Steps
+	err := wait.ExponentialBackoff(retry, func() (bool, error) {
+		err := n.configureDynamically(pcfg)
+		if err == nil {
+			klog.V(2).Infof("Dynamic reconfiguration succeeded.")
+			return true, nil
+		}
+		retriesRemaining--
+		if retriesRemaining > 0 {
+			klog.Warningf("Dynamic reconfiguration failed (retrying; %d retries left): %v", retriesRemaining, err)
+			return false, nil
+		}
+		klog.Warningf("Dynamic reconfiguration failed: %v", err)
+		return false, err
+	})
+	if err != nil {
+		klog.Errorf("Unexpected failure reconfiguring NGINX:\n%v", err)
+		return err
+	}
+
+	ri := utilingress.GetRemovedIngresses(n.runningConfig, pcfg)
+	re := utilingress.GetRemovedHosts(n.runningConfig, pcfg)
+	rc := utilingress.GetRemovedCertificateSerialNumbers(n.runningConfig, pcfg)
+	n.metricCollector.RemoveMetrics(ri, re, rc)
+
+	n.runningConfig = pcfg
+
+	return nil
+}
+
+// OnUpdate is called by the synchronization loop whenever configuration
+// changes were detected. The received backend Configuration is merged with the
+// configuration ConfigMap before generating the final configuration file.
+// Returns nil in case the backend was successfully reloaded.
+func (n *NGINXConfigurer) OnUpdate(ingressCfg ingress.Configuration) error {
+	cfg := n.GetBackendConfiguration()
+	cfg.Resolver = n.resolver
+
+	content, err := n.generateTemplate(cfg, ingressCfg)
+	if err != nil {
+		return err
+	}
+
+	err = createOpentracingCfg(cfg)
+	if err != nil {
+		return err
+	}
+
+	err = n.testTemplate(content)
+	if err != nil {
+		return err
+	}
+
+	if klog.V(2).Enabled() {
+		src, _ := os.ReadFile(cfgPath)
+		if !bytes.Equal(src, content) {
+			tmpfile, err := os.CreateTemp("", "new-nginx-cfg")
+			if err != nil {
+				return err
+			}
+			defer tmpfile.Close()
+			err = os.WriteFile(tmpfile.Name(), content, file.ReadWriteByUser)
+			if err != nil {
+				return err
+			}
+
+			diffOutput, err := exec.Command("diff", "-I", "'# Configuration.*'", "-u", cfgPath, tmpfile.Name()).CombinedOutput()
```

Comment:
I don't like this way. This will generate some IO consumption, we need to consider other more efficient ways